### ISSUE TITLE
The Flink GafferSink add elements queue could run out of memory

### ISSUE THREAD

**Issue by gaffer01** [gchq/Gaffer#1470]
The implementation of `AddElementsFromKafkaHandler` that uses Flink to add data from Kafka contains some code that looks unnecessarily complicated and there's no Javadoc to explain why it's done this way:

- The `invoke` method on `GafferSink` receives an `Iterable` of `Element`s. This will normally be a small batch of `Element`s created from a single Kafka entry. Presumably this is passed to a `GafferSink` which passes it to a `GafferAdder` to avoid the expense of creating an `AddElements` operation for each small batch. If a `flatMap` was called after the `map` in `doOperation` in `AddElementsFromKafkaHandler` then the sink would receive a never-ending iterable of elements which could be passed to a single `AddElement`s operation - only one of these would ever need to be created. There would be no need for a `GafferAdder`.
- The `GafferAdder` maintains a `ConcurrentLinkedQueue` of elements. It looks like it's assumed that elements will be continually added to the queue whilst the `AddElements` operation is simultaneously running and pulling data from it. This means that if the `AddElements` operation is slower than the ingest from Kafka (e.g. because Accumulo is under load) then it will eventually run out of memory, i.e. there is no way of back pressure being applied to earlier functions in the stream. There should be a test to check that back pressure is applied.

If the design is correct then there should be some Javadoc to explain why it's done this way.

Marked as a possible bug in case the OOM issue is true.

**Comment by p013570** [comment]
@gaffer01 - I switched to using a flatMap function to reduce the nested iterable. However, flink calls the sink with each individual element, whether or not I use an iterable in the flatMap. There doesn't seem to be a way around that.

I've changed the queue of elements we give to AddElements to be a blocking queue with a maximum customisable size (defaults to 1,000,000). If the queue is full then flink will be blocked from adding more elements. This should fix the back pressure issue.
